### PR TITLE
Feature Pattern Refinement

### PR DESCRIPTION
**Pull request by lharzenetter**
This feature enables the semi-automatic refinement of Pattern-based Deplyoment Models

## New element types: `Pattern Refinement Models`

image

image

## Refining a Pattern-based Deployment Model

image

Hover over the PRM candidates to see what will be refined:
image

- [x] Ensure that you followed http://eclipse.github.io/winery/dev/ToolChain#github---prepare-pull-request. Especially, we require **a single commit**
- [ ] Ensure that the commit message is [a good commit message](https://github.com/joelparkerhenderson/git_commit_message)
- [x] Ensure to use auto format in **all** files
- [x] Ensure that you appear in `NOTICE` at Copyright Holders
- [x] Tests created for changes
- [ ] Documentation updated
- [x] Screenshots added (for UI changes)

### CODE REVIEW

**Comment by lharzenetter** [comment]
@ckrieger, please read through the code :)

File: org.eclipse.winery.repository.ui/src/app/instance/patternRefinementModels/relationshipMappings/relationMappings.component.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    private handleData(data: any) {
+        this.loading = false;
+        this.relationshipMappings = data[0];
+        this.detectorNodeTemplates = data[1];
+        this.refinementStructureNodeTemplates = data[2];
+        this.relationshipTypes = data[3];
+        this.nodeTypes = data[4];
+        console.log(data);
```
Comment (ckrieger) [inline org.eclipse.winery.repository.ui/src/app/instance/patternRefinementModels/relationshipMappings/relationMappings.component.ts]: console.log

File: org.eclipse.winery.repository.ui/src/app/instance/patternRefinementModels/relationshipMappings/relationMappings.component.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            const number = Number(value.id.split('mapping')[1]);
+            if (!isNaN(number) && number >= id) {
+                id = number;
+                if (number === id) {
+                    id++;
+                }
+            }
+        });
+
+        this.mapping = {
```
Comment (ckrieger) [inline org.eclipse.winery.repository.ui/src/app/instance/patternRefinementModels/relationshipMappings/relationMappings.component.ts]: this.mapping = new RelationMappings() may imporve readbility?

File: org.eclipse.winery.repository.ui/src/app/instance/patternRefinementModels/relationshipMappings/relationMappings.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ *******************************************************************************/
+
+export enum RelationDirection {
+    INGOING = 'INGOING',
+    OUTGOING = 'OUTGOING'
+}
+
+export interface RelationMappings {
```
Comment (ckrieger) [inline org.eclipse.winery.repository.ui/src/app/instance/patternRefinementModels/relationshipMappings/relationMappings.ts]: why RelationMappings isn't it just a single RelationMapping?

File: org.eclipse.winery.repository.ui/src/app/instance/sharedComponents/topologyTemplate/topologyTemplate.component.ts [hunk 4]
```diff
@@ -16,45 +16,52 @@ import { InstanceService } from '../../instance.service';
...
     editorUrl: string;
     oldEditorUrl: string;
+    refinementAvailable = false;
 
     selectedVersion: WineryVersion;
 
     @ViewChild('compareToModal') compareToModal: ModalDirective;
+    compareToModalRef: BsModalRef;
+
+    refinementIsRunning: boolean;
```
Comment (ckrieger) [inline org.eclipse.winery.repository.ui/src/app/instance/sharedComponents/topologyTemplate/topologyTemplate.component.ts]: enum refinementstatus may improve readability :)

Comment (lharzenetter) [reply org.eclipse.winery.repository.ui/src/app/instance/sharedComponents/topologyTemplate/topologyTemplate.component.ts]: I tried, but it just did not work... I had errors when using the enum in the template by using the way of creating a special variable for the enum to reference in the template.

**Pull request merged**